lib/auth/autoupdate/autoupdatev1/service.go

```diff
@@ -648,6 +652,143 @@ func (s *Service) DeleteAutoUpdateAgentRollout(ctx context.Context, req *autoupd
 	return &emptypb.Empty{}, nil
 }
 
+// TriggerAutoUpdateAgentGroup triggers automatic updates for one or many groups
+// in the rollout.
+func (s *Service) TriggerAutoUpdateAgentGroup(ctx context.Context, req *autoupdate.TriggerAutoUpdateAgentGroupRequest) (*autoupdate.AutoUpdateAgentRollout, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.CheckAccessToKind(types.KindAutoUpdateAgentRollout, types.VerbUpdate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.AuthorizeAdminActionAllowReusedMFA(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tries := 0
+	const maxTries = 3
+
+	for {
+		tries++
+		existingRollout, err := s.backend.GetAutoUpdateAgentRollout(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		err = rollout.TriggerGroups(existingRollout, req.Groups, req.DesiredState, s.clock.Now())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		newRollout, err := s.backend.UpdateAutoUpdateAgentRollout(ctx, existingRollout)
+		if err == nil {
+			return newRollout, nil
+		}
+		if !trace.IsCompareFailed(err) {
+			return nil, trace.Wrap(err)
+		}
+		if tries > maxTries {
+			return nil, trace.Wrap(err, "max tries exceeded")
+		}
+	}
+}
+
+// ForceAutoUpdateAgentGroup forces one or many groups of the agent rollout into
+// a DONE state.
+func (s *Service) ForceAutoUpdateAgentGroup(ctx context.Context, req *autoupdate.ForceAutoUpdateAgentGroupRequest) (*autoupdate.AutoUpdateAgentRollout, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.CheckAccessToKind(types.KindAutoUpdateAgentRollout, types.VerbUpdate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.AuthorizeAdminActionAllowReusedMFA(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tries := 0
+	const maxTries = 3
+
+	for {
```

Comment:
Curious about the motivation for repeatedly retrying on conflict here. Is there something we could watch instead?

(Not blocking, just want to understand the architecture better.)

Reply:
I understand the intent behind optimistic locking, but it's unusual to me to see it applied with an arbitrary retry count directly on the backend, vs., e.g., retrying a transaction. I'm assuming we don't have a transaction-like concept given the multitude of backends, and the optimistic lock is our only tool? (Just want to understand this better, not suggesting a change)